src/src/PluginStructs/P145_data_struct.cpp

```diff
@@ -0,0 +1,655 @@
+  }
+};
+const int nbrOfTypes = (int)(sizeof(sensorData) / sizeof(struct P145_SENSORDEF));
+
+/*****************************************************************************/
+/*!
+@brief  Perform the calibration algorithm
+@param[in] currentRcal  Value for Rzero determined with current measurement
+@return Void
+@note   Collect the highest Rzero during the calibration period
+        At the end of the period take mean value of current and calculated Rzero
+*/
+/*****************************************************************************/
+void P145_data_struct::calibrate (float currentRcal)
+{
+  unsigned long now = millis();
+  unsigned long time = now - last_cal;
+  float lastRcal = rcal;  // Last calculated Rcal this calibration sequence
+  bool doit = false;
+  
+  if (currentRcal > rcal)
+  {
+    rcal = currentRcal;
+  }
+  doit = (time > P145_CALIBRATION_INTERVAL) && (rcal > 0.0);
+  if (doit)
+  {
+    rzero = rcal;     // Update Rzero as determined by calibration
+    rcal = 0.0;       // Restart calibration cycle with no value
+    last_cal = now;   // Remember calibration moment
+  }
+
+  if (loglevelActiveFor(LOG_LEVEL_INFO))
+  {
+    String log;
+#ifdef P145_DEBUG
+    log = F("MQ-xx: Calibration with Rcal =  ");
+    log += currentRcal; // Calculated Rzero if calibration concentration is applied
+    log += F(" Rlast = ");
+    log += lastRcal;    // Rcal as calculated previous sample
+    addLog(LOG_LEVEL_INFO, log);
+#endif
+    if (doit)
+    {
+      log = F("MQ-xx: ***Calibrating*** Rzero =  ");
+      log += rcal;
+      addLog(LOG_LEVEL_INFO, log);
+    }
+  }
+}
+
+/*****************************************************************************/
+/*!
+@brief  Get the resistance of the sensor, ie. the property to be measured
+@param[in] val     Analog input value read from ADC [counts]
+@return The sensor resistance in Ohm
+@note   Uses hard coded constants:
+        MAX_ADC_VALUE Max range for ADC
+        VMAX          Analog  input voltage corresponding to MAX_ADC_VALUE
+        VCC           Voltage applied to the sensor-Rload combination
+*/
+/*****************************************************************************/
+float P145_data_struct::getResistance(float val)
+{
+  return ((MAX_ADC_VALUE * P145_VCC) / (P145_VMAX * val) - 1.0) * rload;
+}
+
+/*****************************************************************************/
+/*!
+@brief  Get the resistance RZero of the sensor for calibration purposes
+@param[in] rSensor    Sensor resistance value Rs
+@return The sensor resistance RZero in kOhm
+*/
+/*****************************************************************************/
+float P145_data_struct::getRZero(float rSensor)
+{
+  float newValue = rSensor;  // Default to current measured Rsensor
+  
+  if (newValue < 0.0) newValue = 0.0;  // Clip negative values
+  switch (algorithm)
+  {
+    case p145AlgA:  // MQ-135
+      newValue = (rSensor * pow((refLevel / sensordef.para), (1.0 / sensordef.parb)));
```

Comment:
please use `powf` here as the compiler would otherwise convert the floats to `double` and use the `double` version of `pow`